### PR TITLE
Широких -- Лабораторная работа #1

### PR DESCRIPTION
**Pull request by ViolettaShirokikh**

### CODE REVIEW

**Comment by ViolettaShirokikh** [comment]
@Selllavi, @Yaroslav-Lahtachev review, please

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    private static final int[] ARABIC_NUMERALS = {1000, 900, 500, 400,
+            100, 90, 50, 40, 10, 9, 5, 4, 1};
+    private static final String[] ROMAN_NUMERALS = {"M", "CM", "D", "CD",
+            "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
+
+    public ArabicToRomanConverter() { };
+
+    public String convert(final int arabicNumber) {
+
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java]: Remove this white space

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicConverter.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.Hashtable;
+
+public class RomanToArabicConverter {
+
+    private static final int[] ARABIC_NUMERALS = {1, 5, 10, 50,
+            100, 500, 1000};
+    private static final Character[] ROMAN_NUMERALS = {'I', 'V', 'X',
+            'L', 'C', 'D', 'M'};
+    private Hashtable<Character, Integer> ht;
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicConverter.java]: what is ht?
Name should be meaningful.

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package ru.unn.agile.ArabicRomanConverter;
+
+public class ArabicToRomanConverter {
+
+    private static final int[] ARABIC_NUMERALS = {1000, 900, 500, 400,
+            100, 90, 50, 40, 10, 9, 5, 4, 1};
+    private static final String[] ROMAN_NUMERALS = {"M", "CM", "D", "CD",
+            "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java]: Do we really need this? CM CD and others? In RomanToArabicConverter we have only 
private static final Character[] ROMAN_NUMERALS = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};
and that's seems enough.

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @Test
+    public void canConvert3() {
+        String romanNumber = convertToRoman(3);
+        assertEquals("III", romanNumber);
+    }
+
+    @Test
+    public void canConvert5() {
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java]: why tests are not in correct order? CanConvert3 -> CanConvert4 -> CanConvert5

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+public class ArabicToRomanTest {
+
+    private String convertToRoman(final int arabic) {
+        ArabicToRomanConverter converter = new ArabicToRomanConverter();
+        return converter.convert(arabic);
+    }
+
+    @Test
+    public void canConvertToEmptyString() {
+        String romanNumber = convertToRoman(0);
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java]: Why convertToRoman(0) should return ""?
Maybe in that case you should throw IllegalArgumentException with message "Roman number doesn't exist for Arabic {}"

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        assertEquals("", romanNumber);
+    }
+
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java]: The same in negative cases when you got -1 for instance, add test on negative int. 
You can use @Test(expected = IllegalArgumentException.class) construction in test to expect exception.

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Test
+    public void canConvertTo444() {
+        int arabic = convertToArabic("CDXLIV");
+        assertEquals(444, arabic);
+    }
+
+    @Test
+    public void canConvertTo3999() {
+        int arabic = convertToArabic("MMMCMXCIX");
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java]: I think you should decide what to use just 'arabic' and 'roman' in variables names, or 'romanNumber ' and 'arabicNumber' in order to bring code to uniformity.

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class RomanToArabicTest {
+
+    private int convertToArabic(final String roman) {
+        RomanToArabicConverter converter = new RomanToArabicConverter();
+        int arabic = converter.convert(roman);
```
Comment (Selllavi) [inline code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java]: what will happen if you call converter.convert("BUGAGASHENKA") ? 
I guess you should add test with value that actually not a roman at all. And check it in convert before converting and throw exception if it isn't roman.

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java (hunk 1 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java]: Done

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicConverter.java (hunk 2 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicConverter.java]: I have renamed it

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java (hunk 4 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java]: Done

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java (hunk 5 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java]: Done

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java (hunk 6 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanTest.java]: Done

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java (hunk 7 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java]: Done

File: code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java (hunk 8 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/test/java/ru/unn/agile/ArabicRomanConverter/RomanToArabicTest.java]: Done

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java (hunk 3 above)
Comment (ViolettaShirokikh) [reply code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java]: In RomanToArabicConverter it's enough, but in this case it's easier to use such set . For example, when I convert 4 to "IV", I don't need to count "I" and replace it  by "IV". I can convert it immediately.

**Comment by ViolettaShirokikh** [comment]
@Selllavi, @Yaroslav-Lahtachev see again, please

File: code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package ru.unn.agile.ArabicRomanConverter;
```
Comment (might52) [inline code/lab1-tdd-arabic-roman-converter/src/main/java/ru/unn/agile/ArabicRomanConverter/ArabicToRomanConverter.java]: Please change the package name to low case.

**Comment by might52** [comment]
@ViolettaShirokikh you need to have 2 stars at least.

**Comment by ViolettaShirokikh** [comment]
@VladimirTyagunov review, please

**Comment by vtyagunov** [comment]
:+1:

**Comment by Yaroslav-Lahtachev** [comment]
You can add the test, where u’ll inter ‘IIII’ roman number for example, and u’ll expect exception.

**Comment by Selllavi** [comment]
Now it is okay.
:+1:
